src/ShallowWrapper.js

```diff
@@ -346,7 +345,7 @@ class ShallowWrapper {
     if (this.root !== this) {
       throw new Error('ShallowWrapper::setState() can only be called on the root');
     }
-    if (this.instance() === null || isFunctionalComponent(this.instance())) {
+    if (this.instance() === null || this.renderer.getNode().nodeType === 'function') {
```

Comment:
We have a`getNodeInternal` helper but it returns the rendered node, and here we need the actual root node.